### PR TITLE
[dagster-dbt] state selection

### PR DESCRIPTION
**Pull request by alangenfeld**
Adds `state_path` (back) to `DbtManifestAssetSelection` to facilitate state selection and puts that under test in combination with using `--defer`.

previously added in https://github.com/dagster-io/dagster/pull/12432 and then removed in https://github.com/dagster-io/dagster/pull/14980. Reborn again here.

## How I Tested These Changes

added test

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19649** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19649?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19649?utm_source=stack-comment).

**Comment by alangenfeld** [comment]
I am not certain we want to add this, but figure the PR is a good place to discuss. 

My main motivation was this example https://github.com/dagster-io/hooli-data-eng-pipelines/blob/master/hooli_data_eng/assets/dbt_assets.py#L163-L184 and hoping that a native asset selection approach would solve some of the problems like partitions not working.

**Comment by OwenKephart** [comment]
@alangenfeld My take is that the partition use case does not work for that particular workaround simply because partition information is not being passed in anywhere. A similar issue would arise if a job that targeted a set of partitioned dbt assets with this selection was fired off without a partition key passed in, so I think the issue exists (and is solvable) independent of this change (noting that state:modified on an incremental model will end up rematerializing all partitions of that model). The key is simply to adjust that op to create an event for all partition keys of that asset.

I would prefer to keep this flag out of the DbtManifestAssetSelection as it would mostly be a source of confusion (imo), as we don't really have a way for users to make this selection do what it would do in the typical slim-CI workflow.

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_state.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        assets=[my_dbt_assets],
+        jobs=[
+            define_asset_job(
+                "dbt_state_modified",
+                selection=DbtManifestAssetSelection.build(
+                    manifest=prod_manifest,
+                    select="state:modified",
+                    state_path=TEST_PROJECT_DIR,
+                ),
+            ),
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/core/test_state.py:32]: [1]

**Review by alangenfeld — COMMENTED** [review]
> as we don't really have a way for users to make this selection do what it would do in the typical slim-CI workflow

Can you expand on this? 

From what I could tell [1] nets equivalent behavior to running `dbt run  --select state:modified --state ../dagster_dbt_test_project` directly which is the what I have seen suggested to run for "slim CI" when using dbt core (plus or minus some tweaks to the exact select criteria). 

I am guessing there is some aspect that I do not yet understand.

**Comment by OwenKephart** [comment]
@alangenfeld The major difference is that the slim-CI workflow compares the new manifest.json file to the most recent manifest.json file from an _executed_ run. Example scenario in which this differs from the behavior of just executing a job with this selection any time you deploy:

Imagine you deploy dbt_project vN, that has a large set of changes compared to the previous version, but alongside those changes, there is a critical error in some other part of dagster land which makes it impossible to execute anything. Now, you re-deploy your code location. The state manifest will be identical to the previous manifest because nothing in the dbt project changed, meaning nothing will be selected by state:modified. However, the last time your dbt project was actually executed, the manifest did have a large set of differences.

In general, any scenario in which the approximation `last_deployed_manifest == last_executed_manifest` is not true is subject to similar issues, and that's a fairly large chunk of situations. You can somewhat get away with that approximation if you have confidence that your CI will always launch a successful run, and no runs will start on your deployment before that run executes, so it's possible to set something up that kinda works, it's just not something we should necessarily advertise as good way of approaching things.

**Comment by alangenfeld** [comment]
Do you think these scenarios prevent this from being useful in the context of shorter lived branch deployments / PRs when they are comparing against a prod manifest? My initial reaction when reading through the scenario you presented was thinking that if that initial change w/ problem --> fix happened in a PR, then this state selection would continue to select all the changed models relative to prod in the fix update. 

I can see how there are still situations where it would be wrong, but it seems to me like it could still be useful for development use cases.

**Comment by alangenfeld** [comment]
Had a follow up conversation out of band. Current stance is that while this is marginally better in some ways than current op based workarounds, that the potential confusion / problematic use / documentation-for-caveats make it net not worth it.

**Comment by alangenfeld** [comment]
back link to https://github.com/dagster-io/dagster/issues/15911

**Pull request closed without merging**